File: cpp/src/Function.cpp
//
// Created by Dylan on 6/4/2021.
//

#include "Function.h"

#include <utility>


void Function::addParameter(std::unique_ptr<FunctionParameter> parameter) {
    this->parameters.push_back(std::move(parameter));
    this->_onChangedSignal.dispatch();
}

Function::Function(std::string name, std::string id)
        : id(std::move(id)), name(std::move(name)) {
}
Function::Function(std::string name)
        : id(Code::generateUuidV4()), name(std::move(name)) {
}

const std::string &Function::getName() const {
    return name;
}

std::vector<FunctionParameter *> Function::getParameters() {
    std::vector<FunctionParameter *> answer;
    for (const auto & parameter : this->parameters) {
        answer.push_back(parameter.get());
    }
    return answer;
}

const std::string &Function::getId() const {
    return id;
}

Node * Function::getRootNode() {
    if (!_rootNode) {
        std::cerr << "Function::getRootNode null" << std::endl;
    }
    return _rootNode.get();
}

void Function::setRootNode(std::unique_ptr<Node> rootNode) {
    rootNode->setParent(std::make_unique<NodeParent>(this));
    _rootNode = std::move(rootNode);
    _onChangedSignal.dispatch();
}

Project *Function::getProject() {
    return _project;
}

void Function::setProject(Project *project) {
    _project = project;
}

Signal * Function::getOnChangedSignal() {
    return &_onChangedSignal;
}

File: js/go/ast/function.go
package ast

import "expressionista/common"

type Function struct {
	common.Name
	rootNode   Node
	parameters []*FunctionParameter
}

func (f *Function) addParameter(s string) *FunctionParameter {
	parameter := &FunctionParameter{}
	parameter.SetName(s)
	parameter.function = f
	f.parameters = append(f.parameters, parameter)
	return parameter
}

func (f Function) setRootNode(node Node) {
	f.rootNode = node
}

func NewFunction(s string) *Function {
	return &Function{
		Name:       common.Name{Name: s},
		rootNode:   nil,
		parameters: make([]*FunctionParameter, 0),
	}
}
